poketechnika/Game.cpp

```diff
@@ -1,81 +1,62 @@
-        Camera::setCameraSize(sf::VideoMode::getDesktopMode().width, sf::VideoMode::getDesktopMode().height);
-    }
-    else Camera::setCameraSize(w, h);
-    rend.setScale();
-}
+    // Create all scenes used in game
+    MenuScene menu_scene{};
+    GameScene game_scene{};
+    scenes = {
+        {State::MAINMENU, std::make_shared<MenuScene>(menu_scene)},
+        {State::GAME, std::make_shared<GameScene>(game_scene)},
+    };
+    // TODO: Create all scenes (battle, error)
 
-void Game::processInput() {
-    im.reset_actions();
-    const auto state = sm.getSceneType();
-    im.processInput(window, state);
+    // Set first scene to menu
+    auto first_scene{ scenes.at(State::MAINMENU) };
+    sm.set_scene(std::move(first_scene));
 
-    static std::map<Action, State> changeStateActions{
-        {Action::ChangeSceneToGame, State::GAME},
-        {Action::ChangeSceneToMenu, State::MAINMENU},
-    };
-    for (const auto& elem : changeStateActions) {
-        if (im.getAction(elem.first)) {
-            gm.unloadMap();
-            sm.changeScene(elem.second);
-            if (elem.second == State::GAME) {
-                gm.loadMap();
-            }
-        }
-    }
-    if (im.getAction(Action::Close)) {
-        gm.unloadMap();
-        window.close();
+    if (fullscreen) {
+        width = sf::VideoMode::getDesktopMode().width;
+        height = sf::VideoMode::getDesktopMode().height;
+        window.create(
+            sf::VideoMode(width, height), "Poketechnika", sf::Style::Fullscreen
+        );
     }
+    Camera::setCameraSize(width, height);
+    sm.set_renderer_scale();
 }
 
-void Game::update() {
-    float time = clock.restart().asMilliseconds();
-    for (auto gameObject : gameObjects) {
-        if (gameObject->isActive()) {
-            gameObject->update(time);
+void Game::process_input() {
+    ih.reset_actions();
+    ih.process_input(window);
+    for (auto& elem : change_scene) {
+        if (ih.get_action(elem.first)) {
+            auto next_scene{ scenes.at(elem.second) };
+            sm.set_scene(std::move(next_scene));
         }
     }
+    if (ih.get_action(Action::Close)) {
+        window.close();
+    }
 }
 
-void Game::initGameLoop() {
-    //Create all necessary instances before game starts
-    GameObject::setGameObjectsPtr(&gameObjects);
-    GameObject::setInputManagerPtr(&im);
-    GameScene::setGMPtr(&gm);
-    gm.start();
-    sm.createFirstScene();
-    //Start the loop
+void Game::run() {
     window.setFramerateLimit(60);
 
-    while (window.isOpen())
-    {
-        //Process Input
-        processInput();
-        //Update
-        update();
-        //Render
-        sm.renderScene();
+    // Game loop
+    while (window.isOpen()) {
+        process_input();
+        float time = clock.restart().asMilliseconds();
+        sm.run_scene(time);
     }
-    //Destroy all the things after the window closes
-}
\ No newline at end of file
+    // TODO: Save all data after window closes
```

Comment:
no nie, save będzie w momencie przejścia gamescene -> menuscene, później wywalamy wyłączanie esc poza menu

Reply:
A co w sytuacji gdy ktoś szybko wyjdzie to menu i zaraz wraca do gry? zapisywanie w takiej sytuacji nie ma sensu. Nie lepiej po prostu przy końcu programu i ewentualnie po wybraniu takiej opcji w menu?